### ISSUE TITLE
Resuming from snapshot and using start_depth

### ISSUE THREAD

**Issue by Ianmcmill** [akanimax/pro_gan_pytorch#15]
I have a question for understanding.
I have set feedback_factor: 10
When I interrupt training at depth 3 (32x32) and want to resume training from the saved model, what is correct?
--start_depth=3 --generator_file=training_runs/dataset/saved_models/GAN_GEN_3.pth
or
--start_depth=3 --generator_file=training_runs/dataset/saved_models/GAN_GEN_4.pth

or do I skip depth 3 and start with depth 4?
Should I only interrupt when the depth is chaning. Say during the transition from depth 3 to 4. So that I don't loose epochs or override epochs? For example when I interrupt during depth 3 with a max epoch size of 64 and I stop training on epoch 30. Is this bad? Should I stop at epoch 1 instead?

**Comment by akanimax** [comment]
Hi @Ianmcmill,

Thank you for your question. Please find the explanation for your queries below:

1.) Feedback factor has nothing to do with actual training. It's just for logging the loss values and generating samples during training.

2.) This is a very good question. The process of resuming is a bit involved: 
My design for resuming is as follows:

    a.) If you interrupted the training during the fade-in phase of two layers, ex. from 3 - 4, you will not 
         have saved files for depth 4, so that is out of the question. However, please note that  in this case 
         you cannot resume from the fade-in interruption and you need to start from the previous depth 
         itself.
         So, use: --start_depth=3 and --generator_file=GAN_GEN_2.pth # note one below the depth

    b.) If you interrupt during the stabilization epochs, it is actually much easy to resume (No 
         computations lost).
         So, use: --start_depth=3 and --generator_file=GAN_GEN_3.pth (Also set the fade-in percentage for 
         this depth to 0 now. And, adjust the remaining epochs accordingly, because fade-in is complete).

Hope this helps. Please let me know if you need any more help.

Best regards,
@akanimax

**Comment by Ianmcmill** [comment]
Thanks for answering @akanimax !
I wasn't aware of the fade-in and stabilization. My fault.
Regarding b)
```
# % of epochs for fading in the new layer
fade_in_percentage:
  - 50 #depth 0
  - 50 #depth 1
  - 50 #depth 2
  - 0 #depth 3 - resume with start_depth=3 - set this to 0
  - 50
  - 50
  - 50
```
Would this be correct?

**Comment by akanimax** [comment]
Hi @Ianmcmill,

Yes that is correct. Also reduce the number of epochs in the config to account for the completed ones. For instance, if you were training for 54 epochs with 50% fade in, and you got interrupted at epoch 39 (after fade-in, i.e. during stabilization), then you need to change the number of epochs to 15 in order to save computations.

Hope this helps.

Best regards,
@akanimax

**Comment by Ianmcmill** [comment]
Ok got it.
One last question.
I haven't understood until yet when fade-in happens and when stabilization happens. When do those processes happen?

**Comment by akanimax** [comment]
Hi @Ianmcmill,

No problem.
The fade in happens for the `fade-in percentage` number of epochs for the particular depth. For instance, at depth 3 (i.e. resolution 32 x 32), your epochs are 100; and your fade-in percentage is 50, then the fade in iterations will happen till 50 epochs and remaining 50 epochs will be stablization epochs.

Hope this helps.

Best regards,
@akanimax

**Comment by Ianmcmill** [comment]
I'd suspected something like this. But I wasn't sure if it would be that easy. 
Thanks for the answer and thanks for sharing your code!

**Comment by Ianmcmill** [comment]
I need to ask nother question.
I just understood the consequences when resuming during a fade-in interruption.
I am at depth 3 fade-in. I train for 256 epochs. My cloud instance reset at epoch 60 during the fade-in. With epoch 128 fade-in would be completed. To train to epoch 60 took 6 hours.
With this current design the 6 hours are wasted and I need to start over from depth=3 and models from depth=2.
This is a bit of drawback of this design and I am not sure if I want to use this design in the future. Is there a workaround for efficient resuming from fade-ins? 
I imagine that a complete training on depth=7 with 256 epochs(on my setup) takes around 2833 hours which is roughly 4 month (assuming a 4.66-times increase during each depth). It's no fun starting over the fade-in training.

**Comment by akanimax** [comment]
Hi @Ianmcmill,

Yes you are right. I believe, there is something that can be done in this regards. Basically, the only values required to resume training from mid fade-in are: 1.) the number of epochs completed and 2.) The exact value of alpha. So if the training code is also check-pointing these values along with weights and losses, modifying the code for resuming from mid fade-in stoppage would be possible. 

One thing that you can do right off is: manually calculate the value of alpha for your fade in so that you can get your model to train quickly.

Thanks for pointing out. I will include this in the development of next release. 
BTW, please feel free to chip in if you would like to add this feature and / or you have a better way of doing it.

Best regards,
@akanimax

**Comment by Ianmcmill** [comment]
I acutally do the alpha calculations and adjust the config that way when resuming. Like this
```
# % of epochs for fading in the new layer
fade_in_percentage:
  - 50
  - 50
  - 50
  - 37.6
  - 50
  - 50
  - 50
```
 The samples generated after resuming, override those samples that were generated from the deeper trained network. This made me realize the problem with resuming from fade-ins.

You say "get your model train quickly". How does such a resume from fade-in impact the up to this date trained network? Will it get up to par with the state at the training interruption fast? Or does it have an unwanted impact of the networks convergence?

I would like to help with adding this feature but I must admit I am more of an enthusiast in NN and my python/pytorch knowledge is limited in these regards. I do try to learn it however.

Sample during fade-in and last before interrupt (depth 3)
gen_3_75_152.png
![gen_3_75_152](https://user-images.githubusercontent.com/10505658/52175833-1e6e9d00-27aa-11e9-90f5-143e15124498.png)
Sample during resume
gen_3_1_95.png
![gen_3_1_95](https://user-images.githubusercontent.com/10505658/52175836-28909b80-27aa-11e9-8875-ede8b1c34668.png)

**Comment by akanimax** [comment]
Hi @Ianmcmill,

> I acutally do the alpha calculations and adjust the config that way when resuming.

This could work as a short term patch, but I will think of a permanent solution for it. Please do try it out and let me know how it works.

> You say "get your model train quickly". How does such a resume from fade-in impact the up to this date trained network? Will it get up to par with the state at the training interruption fast? Or does it have an unwanted impact of the networks convergence?

Well, if the resuming process is able to completely match the state of interruption, there should be no effect on the network performance (which is something I will try to do). At the moment, if you resume from a completely faded-in layer or a new layer at higher depth altogether, there is absolutely no effect on the performance.

Thanks for sharing the samples; they look pretty cool given that they are just 32 x 32 images.
I don't quite understand what you mean by:

> The samples generated after resuming, override those samples that were generated from the deeper trained network. This made me realize the problem with resuming from fade-ins.

Basically, if you have trained your model perfectly till depth 7 i.e. 256 x 256 and then later on if you use this model to generate samples from the layer at 32 x 32 (any previous layer), then pro-gan doesn't generate your required images (it creates a structured noise similar to your samples). I believe you have checked out my `MSG-GAN`. There you will always get required images at all resolutions.

Hope this helps.

Best regards,
@akanimax

**Comment by Ianmcmill** [comment]
>The samples generated after resuming, override those samples that were generated from the deeper trained network. 

- Feedback generation:

What I mean is the script generates feedback images(gen_3_1.png, etc). When resume training the script generates feedback images again (gen_3_1.png) effectly overwriting the former genereted feedback images. This is not really an issue but if you want to have a continous feedback stream of images this is not working. And if you want to discuss the model like I do, I have to sum the filename numbers (After interrupting at epoch 60 and resuming training to epoch 63 is effectivly training to epoch 123). This gets overly complicated and confusing when trying to explain an issue :)
However this is something I could try to modify the feedback code. Just insert a check for existing files and name the new feedback accordingly. This is not the real issue.

- Resume training and overwriting model at interruped state:

This is difficult to explain. Danger: lots of images incoming!
Config: 
epochs=256
fade-in %=50
Fade-in was completed with epoch 128.
Epoch 174 is stabilizing.
Training from depth 3 epoch 0 to depth 3 epoch 174, this is the feedback from d3_e174:

![gen_3_114_152](https://user-images.githubusercontent.com/10505658/52203458-8cca6280-2871-11e9-891d-d1e75ba27371.png)
--------------------------------------------------
This is the feedback from d3_e11 after resuming from d3_e174.
I do not have the d3_e1 feedback as this has already been overwritten after the most recent interrupt.
Config:
epochs=82 (remaining epochs to stabilize until an overall of 256 epochs have been trained)
fade-in %=0
![gen_3_11_19](https://user-images.githubusercontent.com/10505658/52203637-10844f00-2872-11e9-9f74-18fd951e24fa.png)
As you can see, those images after the resume are foggy. 

--------------------------

Same here. Stabilizing training was interrupted at d3_e230.
Last feedback before interrupt (Colab Instance reset)
![gen_3_73_19](https://user-images.githubusercontent.com/10505658/52203666-22fe8880-2872-11e9-9145-8dc6106ccad4.png)

And after resuming from d3_e230 this is the feedback of the first epoch
Config:
epochs=26
fade-in %=0
![gen_3_1_1](https://user-images.githubusercontent.com/10505658/52203703-3d386680-2872-11e9-8f40-5e68d96fb772.png)

and this is the feedback after 10 more epochs:
![gen_3_11_1](https://user-images.githubusercontent.com/10505658/52203727-4c1f1900-2872-11e9-9a55-a71a4953ff6e.png)


As you can see, after resuming it takes some epochs (15-20, depening on the dataset size and hyperparameters) to recover from the random noise that the first feedback after resuming training happens. those blue-ish feedback image. I wonder how this process of recovering affects the network. The feedback images after a resume and after recovering show a visible shroud/haze. The colors are washed out and some sharp details are gone.

**Comment by Ianmcmill** [comment]
@akanimax so what does the network actually do when I resume training from an interrupted fade-in training?
I reached depth 4 and trained fade-in until epoch 36 which took 7 hours. Then the instance reseted.
I should now start again with `python train.py --start_depth=4 --gen=gan_gen_3.pth` (simplified).
What happens to the model when I resume with `python train.py --start_depth=4 --gen=gan_gen_4.pth` instead and modify the config as follows. Will it fade-in depth 4 with itself and therefore not actually fade-in at all?:
```
# Training hyperparameters:
epochs:
  - 128
  - 256
  - 256
  - 256
  - 220 ( 256 epochs - 36 completed fade-in epochs; 92 fade-in epochs remaining)
  - 256
  - 256

# % of epochs for fading in the new layer
fade_in_percentage:
  - 50
  - 50
  - 50
  - 50
  - 41.8 ( 41.8% == 92 epochs from 220 overall epochs)
  - 50
  - 50
```
Having just a single GPU at hands and even a limited Tesla K80 is a real setback when experimenting with PPGANs. I cannot let my desktop GPU run all the time as I need to work on my machine too. And although a Google VM with free GPU and 12gb VRAM is nice, the 12h (at most) reset timer doesn't help in this regard as well.
Maybe I should set up a Google Cloud VM instance and use my free credit I get for trail. A persistant V100 instance is 1.8$/h but a K80 should be about half. With a credit budget of 300$ free this should be able to run a full month :)

**Comment by Ianmcmill** [comment]
I've deployed the model to a persistent Goolge Cloud Compute Engine. I'll report back in four days with the results.

**Comment by Ianmcmill** [comment]
> a.) If you interrupted the training during the fade-in phase of two layers, ex. from 3 - 4, you will not 
>      have saved files for depth 4, so that is out of the question. However, please note that  in this case 
>      you cannot resume from the fade-in interruption and you need to start from the previous depth 
>      itself.
>      So, use: --start_depth=3 and --generator_file=GAN_GEN_2.pth # note one below the depth

I have to ask this again. I don't get it.
Why do I have to start from depth 3 with model from depth 2 again when I resume at the beginning of fade-in from depth 3 to 4?
When depth 3 training is completed and depth 4 starts to train, there is a completly trained gan_gen_3.pth.
Why do I need to start again from gan_gen_2.pth. This would then start training from the model from depth 2. This means it would start to snapshot depth 3 models.

As you said in the quote, if I interrupt training at 
```
Currently working on Depth:  5
Current resolution: 128 x 128
Epoch: 1
Elapsed: [2 days, 0:00:31.778186]  batch: 1  d_loss: -16.960335  g_loss: 28.111532
Elapsed: [2 days, 0:03:36.965419]  batch: 22  d_loss: -34.446026  g_loss: 38.343319
Elapsed: [2 days, 0:06:48.461571]  batch: 44  d_loss: -25.944784  g_loss: 69.050522
Elapsed: [2 days, 0:09:59.920726]  batch: 66  d_loss: -40.334137  g_loss: 58.468468
Elapsed: [2 days, 0:13:11.610885]  batch: 88  d_loss: -49.160400  g_loss: 90.675735
Elapsed: [2 days, 0:16:23.425243]  batch: 110  d_loss: -48.781715  g_loss: 75.740211
Elapsed: [2 days, 0:19:34.784112]  batch: 132  d_loss: -46.364792  g_loss: 74.733574
Elapsed: [2 days, 0:22:46.124331]  batch: 154  d_loss: -45.037827  g_loss: 53.111076
Elapsed: [2 days, 0:25:57.664086]  batch: 176  d_loss: -29.315836  g_loss: 66.617813
interruped...
```
I would then need to resume with
`python train.py --start_depth=4 --generator=gan_gen_3.pth`
This would render 48h hours of training a complete depth (fade-in + stab) obsolete.

From common sense I would resume training with
`python train.py --start_depth=5 --generator=gan_gen_4.pth`
gan_gen_4.pth is a fully completed model snapshot of depth 4 (256 epochs).
Am I getting this wrong or is the way it resumes highly inefficient compared to tkarras progan?

Also:
This is an image from an uninterrupted training after completing depth 4 (all 256 epochs trained) and starting with depth 5 (fade-in)
![gen_5_1_1197](https://user-images.githubusercontent.com/10505658/52410643-437c4c00-2ad9-11e9-8e16-54b240f666ae.png)
gen_5_1_1.png (not interrupted)
-------------
This is an image after an interrupt directly after completing depth 4, starting with depth 5 and resuming with 
`python train.py --start_depth=5 --generator=gan_gen_4.pth`
![gen_5_1_1](https://user-images.githubusercontent.com/10505658/52410630-37908a00-2ad9-11e9-89cf-a0d4c09e21a4.png)
gen_5_1_1.png (resume after interrupt)

**Comment by Mut1nyJD** [comment]
Yes I am seeing similar results.  Not sure why this issue was closed.

Resuming a training session seems pretty much impossible if you do not want to end up in an unstable situation.

But a few questions to you not also have to load the discriminator state as well, I don't see that in your command line you only seem to load the generator that seems wrong.  
What I do not understand though do I have to load the shadow file or the generator or both this is not quiet clear to me. 

But even if I do load both discriminator and generator it certainly does not pick up where it was left. If you have just completed a depth and restart at the next depth using the gen/dis files from the previous depth does not lead to the same results as when you interrupted that depth it it degrades into noise.

Another question why do you need these crazy amount of epochs?

**Comment by akanimax** [comment]
Well, I understand that I have not described the resuming process properly. I know this is not the best solution, but then again finding a very robust solution for resuming from a training interruption is not very straightforward. I have pointed this out in our new paper -> [https://arxiv.org/abs/1903.06048](https://arxiv.org/abs/1903.06048). 

Anyway my design of the resuming process is:
1.) If interrupted during stabilization iterations, just change the number of epochs in the config and resume directly
2.) if interrupted during fade-in iterations, resume from a previously stabilized depth (resolution).
And, in order to resume, you need to provide all 5 weight files, i.e. GAN_GEN, GAN_GEN_SHADOW, GAN_DIS, GAN_GEN_OPTIM, GAN_DIS_OPTIM. 

> Another question why do you need these crazy amount of epochs?

True, in order to correctly get results, you have to spend these many training iterations at every resolution. Otherwise you get blobby and wavy results. I have pointed this out in the paper as well.

Hope this helps. 

Best regards,
@akanimax

**Comment by Mut1nyJD** [comment]
Thanks I do not expect it to be simple that is the problem with progressive training.  :)
 
Your paper looks interesting.

I assumed though if you wait for one level to be finished and interrupt at that point it should not to be too hard to restart from that.  

Thank you for clearing up that you need to load all of these this was not clear to me and was not really reflected well in any of the answers above.

About the epochs I was not saying in general I was referencing to the numbers from Ianmcmill  (I do understand that) but something like 128 epochs sounds extremely high to me and you have to be really careful on what GAN strategy you are using because a lot of them tend to become instable after too many epochs..
